### PR TITLE
[cleanup] Replace atomic functions with std::atomic

### PR DESCRIPTION
**Pull request by Rechi**
## Description
Replace atomic functions with `std::atomic` functions.
~~Replace atomic functions with intrinsics.
Cherry picked witch permission from @Paxxi.~~

## Motivation and Context
Cleanup of architecture dependent code  and inline assembler isn't supported on Windows x64.

## How Has This Been Tested?

Compiled on Windows, OS X and Linux.

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by FernetMenta** [comment]
why don't you trash this outdated functions completely? with move to c++11 those are not needed anymore.

**Comment by Rechi** [comment]
@FernetMenta updated
Do you mean something like that? I think the function calls should just replaced with the body of the function.
`cas` still needs to be updated to the new style.

File: xbmc/interfaces/legacy/AddonClass.h [hunk 1]
```diff
@@ -126,7 +126,7 @@ namespace XBMCAddon
     void Release() const
 #ifndef XBMC_ADDON_DEBUG_MEMORY
     {
-      long ct = AtomicDecrement((long*)&refs);
+      long ct = AtomicDecrement((std::atomic<long>*)&refs);
```
Comment (FernetMenta) [inline xbmc/interfaces/legacy/AddonClass.h]: AtomicDecrement can be dropped, operations on atomic declared parameters are atomic

File: xbmc/peripherals/devices/PeripheralImon.h [hunk 2]
```diff
@@ -20,6 +20,7 @@
  */
 
 #include "PeripheralHID.h"
+#include "threads/Atomics.h"
```
Comment (FernetMenta) [inline xbmc/peripherals/devices/PeripheralImon.h]: trash Atomic.h, include <atomic>

**Comment by FernetMenta** [comment]
Test cases are not needed anymore and can be dropped as well

**Comment by Rechi** [comment]
@FernetMenta now I have also dropped `cas` and the Test cases. Hopefully I have done `CAtomicSpinLock` right.

**Comment by Paxxi** [comment]
@Rechi can you send me an email at paxxi at kodi dot tv? It's not related to this pr

**Comment by garbear** [comment]
is `virtual` still needed on std::atomic?

**Comment by FernetMenta** [comment]
@garbear you mean volatile not virtual, right? yes this should be dropped.

@Rechi apart from this it looks good to me.

**Comment by Rechi** [comment]
@FernetMenta I've removed `volatile`.

**Comment by FernetMenta** [comment]
thanks
jenkins build this please

**Comment by MartijnKaijser** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
@Rechi @popcornmix there are build error on Pi: http://jenkins.kodi.tv/job/LINUX-RBPI/9092/console

**Comment by Rechi** [comment]
error should be fixed

**Comment by MartijnKaijser** [comment]
Jenkins build this please

**Pull request merged**